tests/unit/s2n_connection_preferred_keyshare_test.c

```diff
@@ -0,0 +1,127 @@
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+        /* Explicity set default_tls13 security_policy */
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "20190801"));
+        EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+        /* Default security policy for TLS1.3 has ecc_preferences in the following order: x25519, p-256, p-384.*/
+        EXPECT_SUCCESS(s2n_connection_set_keyshare_by_name_for_testing(conn, "x25519"));
+        /* lsb is not set */
+        EXPECT_FALSE(S2N_GENERATE_EMPTY_KEY_SHARE_LIST(conn->preferred_key_shares));
+        /* Bitmap value for x25519 set is 00000010 */
+        EXPECT_TRUE(S2N_GENERATE_KEY_SHARE_FOR_SELECTED_GROUP(conn->preferred_key_shares, 1));
+        EXPECT_SUCCESS(s2n_connection_set_keyshare_by_name_for_testing(conn, "secp384r1"));
+        /* Bitmap value for p-384 and x25519 set is 00001010 */
+        EXPECT_FALSE(S2N_GENERATE_KEY_SHARE_FOR_SELECTED_GROUP(conn->preferred_key_shares, 2));
+        EXPECT_TRUE(S2N_GENERATE_KEY_SHARE_FOR_SELECTED_GROUP(conn->preferred_key_shares, 3));
+        EXPECT_SUCCESS(s2n_config_free(config));
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+        EXPECT_SUCCESS(s2n_disable_tls13());
+    }
+    {
+        /* Test sending duplicate keyshares */
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_SUCCESS(s2n_connection_set_keyshare_by_name_for_testing(conn, "secp256r1"));
+        /* lsb is not set */
+        EXPECT_FALSE(S2N_GENERATE_EMPTY_KEY_SHARE_LIST(conn->preferred_key_shares));
+        /* Bitmap value for p-256 set is 00000010 */
+        EXPECT_TRUE(S2N_GENERATE_KEY_SHARE_FOR_SELECTED_GROUP(conn->preferred_key_shares, 1));
+        EXPECT_SUCCESS(s2n_connection_set_keyshare_by_name_for_testing(conn, "secp256r1"));
+        EXPECT_TRUE(S2N_GENERATE_KEY_SHARE_FOR_SELECTED_GROUP(conn->preferred_key_shares, 1));
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+    /* Test sending keyshares for all curves in default security policy->ecc_preferences: p-256, p-384*/
+    {
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_TRUE(!(conn->preferred_key_shares | 0));
+
+        EXPECT_SUCCESS(s2n_connection_set_keyshare_by_name_for_testing(conn, "secp256r1"));
+        /* lsb is not set */
+        EXPECT_FALSE(S2N_GENERATE_EMPTY_KEY_SHARE_LIST(conn->preferred_key_shares));
+        /* Bitmap value for p-256 set is 00000010 */
+        EXPECT_TRUE(S2N_GENERATE_KEY_SHARE_FOR_SELECTED_GROUP(conn->preferred_key_shares, 1));
+        EXPECT_SUCCESS(s2n_connection_set_keyshare_by_name_for_testing(conn, "secp384r1"));
+        /* Bitmap value for p-256 and p-384 set is 00000110 */
+        EXPECT_TRUE(S2N_GENERATE_KEY_SHARE_FOR_SELECTED_GROUP(conn->preferred_key_shares, 2));
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+    /* Test sending keyshare for curve not supported: Curve x448 */
```

Comment:
Can we test sending a keyshare for a curve not in the security policy?

Reply:
Curve x448 is not in the security policy that is being sent, do you mean supported by s2n and not in the security policy?